x/chainlink/client/rest/rest.go

```diff
@@ -18,10 +28,86 @@ func RegisterRoutes(clientCtx client.Context, r *mux.Router) {
 	registerTxHandlers(clientCtx, r)
 }
 
-func registerQueryRoutes(clientCtx client.Context, r *mux.Router) {
-	r.HandleFunc("chainlink/feed/data/{feedId}", listFeedDataHandler(clientCtx)).Methods(MethodGet)
+func registerQueryRoutes(clientCtx client.Context, rtr *mux.Router) {
+	r := clientrest.WithHTTPDeprecationHeaders(rtr)
+
+	r.HandleFunc("/chainlink/feed/data/round/{roundId}/{feedId}", listRoundFeedDataHandler(clientCtx)).Methods(MethodGet) // query feed data by roundId and feedId
+	r.HandleFunc("/chainlink/feed/data/latest/{feedId}", listLatestFeedDataHandler(clientCtx)).Methods(MethodGet)         // query the latest feed data by feedId
+}
+
+func registerTxHandlers(clientCtx client.Context, rtr *mux.Router) {
+	r := clientrest.WithHTTPDeprecationHeaders(rtr)
+
+	r.HandleFunc("/txs/encode", authrest.EncodeTxRequestHandlerFn(clientCtx)).Methods(MethodPOST)
+	r.HandleFunc("/txs/decode", authrest.DecodeTxRequestHandlerFn(clientCtx)).Methods(MethodPOST)
+	r.HandleFunc("/txs", authrest.BroadcastTxRequest(clientCtx)).Methods(MethodPOST)
+
+	r.Handle("/chainlink/feed/data", submitFeedDataHandler(clientCtx)).Methods(MethodPUT)
+}
+
+func listRoundFeedDataHandler(clientCtx client.Context) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+
+		roundId := vars["roundId"]
+		roundIdInt, err := strconv.ParseInt(roundId, 10, 16)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusBadRequest, errors.New("roundId is invalid").Error())
+			return
+		}
+		feedId := vars["feedId"]
+
+		res, height, err := clientCtx.QueryWithData(fmt.Sprintf("custom/%s/%s/%d/%s", types.StoreKey, types.QueryRoundFeedData, roundIdInt, feedId), nil)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusNotFound, err.Error())
+			return
+		}
+
+		clientCtx = clientCtx.WithHeight(height)
+		rest.PostProcessResponse(w, clientCtx, res)
+	}
 }
+func listLatestFeedDataHandler(clientCtx client.Context) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+		feedId := vars["feedId"]
+
+		res, height, err := clientCtx.QueryWithData(fmt.Sprintf("custom/%s/%s/%s", types.StoreKey, types.QueryLatestFeedData, feedId), nil)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusNotFound, err.Error())
+			return
+		}
+
+		clientCtx = clientCtx.WithHeight(height)
+		rest.PostProcessResponse(w, clientCtx, res)
+	}
+}
+
+func submitFeedDataHandler(clientCtx client.Context) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var req createFeedRequest
+		if !rest.ReadRESTReq(w, r, clientCtx.LegacyAmino, &req) {
+			rest.WriteErrorResponse(w, http.StatusBadRequest, "failed to parse request")
+			return
+		}
+		baseReq := req.BaseReq.Sanitize()
+		if !baseReq.ValidateBasic(w) {
+			return
+		}
+
+		submitter, err := sdk.AccAddressFromBech32(baseReq.From)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
+			return
+		}
+
```

Comment:
do we need validation with the submitter here?
```
if (submitter != req.Submitter) { return someErrorMsg }
```

Reply:
or something along those lines... we can just keep as TODO but would be good to note that this should be done eventually!

Reply:
what you mean by `validation` here? you mean validate signature against pubkey? or rawTx against signature? `ValidateBasic` func will take care of basic input validation, so we dont need to validate in http handler func.

Reply:
no no, just to make sure that the submitter that is being taken from `baseReq.From` is the same as the one in `req.Submitter`

Reply:
hmm.... not sure if that's necessary, `baseReq` is the one to take care of the general cmd flags, if doesnt matter even if they are different, because the payer is always baseReq.From. but just to prevent from this to happen, I will remove `submitter` from `createFeedRequest` so there will be only one field which is `req.baseReq.From`

Reply:
yeah, i thought you were storing for validation but if it isn't used, then we should just remove the variable from the struct. cool 👍